### PR TITLE
feat: std::Futures

### LINKED ISSUE #63: Migrating to futures 0.3
With async/await rolling out into the entire ecosystem, it's time to upgrade existing code to the latest future and to use `std::future`, as that has been stabilized too. I can help out in this too (I have some good amount of free time next weekend)

### PR DESCRIPTION
**Pull request by jhgg**
this ports cassandra_cpp to use `std::futures::Future` which means no more adapters. 

commits in this PR:
 - the big one the first updates to edition 2018 (just uses cargo fix, and then cargo fmt (i just deleted `rustfmt.toml` as a lot of the options there are no longer present in rustfmt :(). Edition 2018 is required for async/await functions.
-  the second one, much smaller (prolly the one to review) ports us to `std::futures::Future` and updates tests to use tokio's testing stuff to run async tests :o 
    - i've implemented `CassFuture.wait()` to just use the c++ blocking future calls, so most code doesn't need to be edited, aside from dropping the now un-needed futures import, as CassFuture provides the wait method. this gets around the need to provide a runtime to drive the future to completion, piggy-backing off of the one already provided in the C++ lib.
    - there is a new function `Cluster.connect_async` which is the non-blocking version of `Cluster.connect`. 

Fixes #63

### CODE REVIEW

**Comment by kw217** [comment]
Wow, thank you - very much appreciated! I'll take a look soon.

In the meantime, please can you modify your commit messages to include the `Signed-off-by` line to confirm you certify the origin of the code, per https://github.com/Metaswitch/cassandra-rs/blob/master/CONTRIBUTING.md ? Thanks.

File: rustfmt.toml [hunk 1]
```diff
@@ -7,7 +7,7 @@ fn_brace_style = "PreferSameLine"
 fn_return_indent = "WithArgs"
 fn_args_paren_newline = true
 fn_args_density = "Compressed"
-fn_args_layout = "Visual"
+fn_args_layout = "Vertical"
```
Comment (kw217) [inline rustfmt.toml]: Changing this has caused very many changes which are just formatting, making this hard to review. I would prefer you left the formatting settings as they were. If you must change the formatting, let's revert completely to the Rust defaults, and put the formatting changes in a separate commit (as you did with the Rust 2018 changes). Thanks.

Comment (jhgg) [reply rustfmt.toml]: done!

File: src/cassandra/cluster.rs [hunk 2]
```diff
@@ -157,11 +156,12 @@ impl Cluster {
     }
 
     /// Performs a blocking call to connect to Cassandra cluster
-    pub fn connect(&mut self) -> Result<Session> {
+    pub async fn connect(&mut self) -> Result<Session> {
```
Comment (jhgg) [inline src/cassandra/cluster.rs]: This is a breaking change - as to avoid having to pick a runtime to drive the future to completion, leaving that as an exercise for the caller (e.g. using tokio's runtime block-on, or just awaiting it...)

File: rustfmt.toml (hunk 1 above)
Comment (jhgg) [reply rustfmt.toml]: For context, this (and many of the other options) don't exist anymore in rustfmt (1.4.9-stable (33e3667 2019-10-07)), so I've just removed them all and re-formatted the project. 

```
Warning: Unknown configuration option `closure_indent_style`
Warning: Unknown configuration option `density`
Warning: Unknown configuration option `enum_trailing_comma`
Warning: Unknown configuration option `expr_indent_style`
Warning: Unknown configuration option `fn_arg_indent`
Warning: Unknown configuration option `fn_args_density`
Warning: Unknown configuration option `fn_args_paren_newline`
Warning: Unknown configuration option `fn_brace_style`
Warning: Unknown configuration option `fn_empty_single_line`
Warning: Unknown configuration option `fn_return_indent`
Warning: Unknown configuration option `generics_indent`
Warning: Unknown configuration option `ideal_width`
Warning: Unknown configuration option `leeway`
Warning: Unknown configuration option `reorder_imported_names`
Warning: Unknown configuration option `rewrite_if_else`
Warning: Unknown configuration option `single_line_if_else`
Warning: Unknown configuration option `struct_lit_style`
Warning: Unknown configuration option `struct_lit_trailing_comma`
Warning: Unknown configuration option `struct_trailing_comma`
Warning: Unknown configuration option `where_density`
Warning: Unknown configuration option `where_indent`
Warning: Unknown configuration option `where_layout`
Warning: Unknown configuration option `where_pred_indent`
Warning: can't set `format_strings = true`, unstable features are only available in nightly channel.
Warning: can't set `fn_single_line = true`, unstable features are only available in nightly channel.
Warning: can't set `report_todo = Always`, unstable features are only available in nightly channel.
Warning: can't set `report_fixme = Never`, unstable features are only available in nightly channel.
```

File: src/cassandra/cluster.rs [hunk 3]
```diff
@@ -161,7 +160,18 @@ impl Cluster {
...
             let session = Session(cass_session_new());
             let connect_future = <CassFuture<()>>::build(cass_session_connect(session.0, self.0));
-            connect_future.wait().map(|_| session)
+            connect_future.wait()?;
+            Ok(session)
+        }
+    }
+
+    /// Asynchronously connects to the cassandra cluster
+    pub async fn connect_async(&mut self) -> Result<Session> {
```
Comment (kw217) [inline src/cassandra/cluster.rs]: There's a bunch of duplicate code between `connect` and `connect_async`. Why not rewrite `connect` to call `connect_async`? If `.wait()` is good enough for `connect` as written, it ought to be good enough to apply to the result of `connect_async`.

File: Cargo.toml [hunk 4]
```diff
@@ -24,4 +24,7 @@ decimal = "1"
 time = "0.1"
 uuid = "0.5"
 error-chain = "0.12.1"
-futures = "0.1"
```
Comment (kw217) [inline Cargo.toml]: Please can you update the CHANGELOG? This will be a major version bump (to 0.15), since it introduces `std::futures`.

I also suspect some clients (including our internal one) will keep using the futures 0.1 version for a while, so we need to keep the 0.14.x stream around in case more changes are required there.

File: tests/futures.rs [hunk 5]
```diff
@@ -36,55 +35,57 @@ fn insert_into_async(
...
         let mut inserts =
             insert_into_async(&session, "test".to_owned(), NUM_CONCURRENT_REQUESTS).unwrap();
         // Put in reverse, so we poll the later ones (which won't be ready) before the earlier ones
         // (which will be immediately ready)
         inserts.reverse();
-
-        create_table
-            .and_then(move |_| use_examples)
-            .and_then(move |_| futures::future::select_all(inserts).map_err(|(e, _, _)| e))
+        futures::future::try_join_all(inserts).await
```
Comment (kw217) [inline tests/futures.rs]: No, this isn't testing the same thing. The purpose of this test (see comment on `test_early_drop_rust_futures`) is to cope with a future being dropped before it is complete. So it's critical that this uses `select_all` not `join_all`. Please can you change it back? (The similar code you've put in `test_rust_futures` is fine, since that's just testing the mainline case.)

File: src/cassandra/future.rs [hunk 6]
```diff
@@ -179,15 +182,22 @@ impl<T: Completable> futures::Future for CassFuture<T> {
...
+            ret
         }
     }
 }
 
+impl<T: Completable> CassFuture<T> {
+    /// Synchronously executes the CassFuture, blocking until it
+    /// completes.
+    pub fn wait(self) -> Result<T> {
+        unsafe { get_completion(self.inner) }
```
Comment (kw217) [inline src/cassandra/future.rs]: This changes the implementation so that the waiting happens purely in C, rather than by a driving loop within `futures`. That's fine, but it relies on knowing that `get_completion` supports such use. It does, because `cass_future_get_error_code` does, but that may not be obvious. Please can you fix the comments on lines 201 and 203 that say "a completed future" and "Future is ready" to reflect the new reality?

File: src/cassandra/future.rs [hunk 7]
```diff
@@ -120,11 +122,10 @@ impl Completable for PreparedStatement {
...
 
 /// A Cassandra future is a normal Rust future.
-impl<T: Completable> futures::Future for CassFuture<T> {
-    type Item = T;
-    type Error = Error;
+impl<T: Completable> Future for CassFuture<T> {
+    type Output = Result<T>;
 
-    fn poll(&mut self) -> futures::Poll<Self::Item, Self::Error> {
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
```
Comment (kw217) [inline src/cassandra/future.rs]: I haven't done any futures migrations myself, but I have skimmed https://www.ncameron.org/blog/migrating-a-crate-from-futures-0-1-to-0-3/ . Can you explain to me why this pin is sufficient and correct?

File: src/cassandra/future.rs [hunk 8]
```diff
@@ -148,29 +149,31 @@ impl<T: Completable> futures::Future for CassFuture<T> {
...
                     }
-                    FutureState::Awaiting { ref mut task, .. } => {
+                    FutureState::Awaiting { ref mut waker, .. } => {
                         // Callback already scheduled; don't set it again (C doesn't support it anyway),
                         // but be sure to swizzle the new task into place. No need to check for
                         // readiness here; we have to wait for the callback anyway so we might as well
                         // do all the work in one place.
-                        *task = futures::task::current();
-                        Ok(futures::Async::NotReady)
+                        if !waker.will_wake(cx.waker()) {
```
Comment (kw217) [inline src/cassandra/future.rs]: This is a change to the existing code. Is it necessary, and why is it correct/safe? The comment says "be sure to swizzle the new task into place", but it looks like we now do *not* necessarily swizzle the new task into place.

(Not saying the old code was necessarily correct either, but it naively looks like the new code differs - though maybe not essentially. Really I'm just looking for an explanation, since I'm not yet familiar with futures 0.3.) Thanks.

File: Cargo.toml (hunk 4 above)
Comment (jhgg) [reply Cargo.toml]: Yup! Will take care of that. Wasn't sure if you wanted that before or after the merge :D

File: src/cassandra/cluster.rs (hunk 3 above)
Comment (jhgg) [reply src/cassandra/cluster.rs]: Yeah, this makes a lot of sense!

File: src/cassandra/future.rs (hunk 8 above)
Comment (jhgg) [reply src/cassandra/future.rs]: Right, so, each time the future is polled, it may or may not get a new waker. This is purely an optimization though. What we are saying is "is the waker that is already in place the same as the one that we were just called with? if it is, then we don't need to do anything, otherwise, if it isn't, we need to swap the waker out." 

So, we're saying "if the waker we gave to C++ land would wake me up, then I have nothing to do, otherwise I have to give it a new waker in order to wake me up." 

This is purely an optimization, and is mentioned in the std docs here: https://doc.rust-lang.org/beta/std/task/struct.Waker.html#method.will_wake

File: tests/futures.rs (hunk 5 above)
Comment (jhgg) [reply tests/futures.rs]: I see it now. Thanks! Will fix.

File: src/cassandra/future.rs (hunk 7 above)
Comment (carllerche) [reply src/cassandra/future.rs]: I'm not sure I understand the question exactly. Switching from `&mut self` -> `self: Pin<&mut Self>` is how one updates manual future implementations (it is a mechanical change).

> why this pin is sufficient and correct?

Why would it not be sufficient and correct? In theory, since `CassFuture` is `!Unpin` there should be no need for `unsafe`.

Comment (carllerche) [reply src/cassandra/future.rs]: As long as all types are `Unpin` (which they should be in a mechanical upgrade from 0.1 -> std::future), then changing `&mut self` -> `self: Pin<&mut Self>` is purely a mechanical change. When types are `Unpin`, `Pin<&mut Self>` is roughly identical to `&mut self`. Things get complicated once you work with `!Unpin` types, i.e. futures returned from `async fn`.

Comment (jhgg) [reply src/cassandra/future.rs]: As long as we are not messing with `!Unpin`, using `Pin` is equivalent to `&mut self`, so this is as safe as it was before :D

Comment (jhgg) [reply src/cassandra/future.rs]: In order to make sure that no one is accidentally messing with `!Unpin`, I've also gone ahead and scoped down the `unsafe` block to wrap the minimal amount of code (the CFFI calls).

**Comment by jhgg** [comment]
@kw217 I've addressed most of your feedback - gonna wait on a reply to https://github.com/Metaswitch/cassandra-rs/pull/65#discussion_r370741975 before I move forward with it though.

File: src/cassandra/cluster.rs (hunk 3 above)
Comment (kw217) [reply src/cassandra/cluster.rs]: OK, doesn't sound worth the engineering. Let's keep this annoying duplication for now.

File: src/cassandra/future.rs (hunk 7 above)
Comment (kw217) [reply src/cassandra/future.rs]: OK, thanks - that makes sense.

File: src/cassandra/future.rs (hunk 8 above)
Comment (kw217) [reply src/cassandra/future.rs]: Cool, that makes sense. I thought that was probably it, but wanted to confirm.

**Comment by jhgg** [comment]
@kw217 sounds good, lets ship 🚀

**Comment by kw217** [comment]
Thanks very much for this fix - it's now released as 0.15.0. Cheers!
